### PR TITLE
Track key combos instead of keys with modifiers

### LINKED ISSUE #1279: Allow using gamepad buttons as positive/negative axis sources

### PR DESCRIPTION
**Pull request by luishfonseca**
# Description

Depends on #1278 

Introduces InputCombinations to replace KeyWithModifiers
Also closes #1279 

Did not test if gamepad was broken with the changes

## Checklist

- [x] Self-review changes.
- [ ] Evaluate impact on the documentation.
- [x] Add entry to the changelog's unreleased section.

### CODE REVIEW

File: core/src/data/ser/debug.cpp [hunk 1]
```diff
@@ -136,6 +138,13 @@ bool DebugSerializer::decompose(const Type& type, const void* value)
         this->separate(true);
         mStream.put(']');
     }
+    else if (type.has<EnumTrait>())
+    {
+        const auto& trait = type.get<EnumTrait>();
+        mStream.put('"');
+        mStream.print(trait.variant(value).name());
+        mStream.put('"');
```
Comment (RiscadoA) [inline core/src/data/ser/debug.cpp:146]:
```suggestion
        mStream.print(trait.variant(value).name());
```
An enum variant is more of a type than a string, I would remove the quotes.

File: engine/include/cubos/engine/input/action.hpp [hunk 2]
```diff
@@ -86,14 +63,12 @@ namespace cubos::engine
         /// @param justPressed Whether this action was just pressed.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/input/action.hpp:63]: Not part of this PR, but pls fix the docs to `Sets whether this action was just pressed.`

File: engine/include/cubos/engine/input/action.hpp [hunk 3]
```diff
@@ -86,14 +63,12 @@ namespace cubos::engine
...
         void justPressed(bool justPressed);
 
-        /// @brief Checks if this accion was just released.
+        /// @brief Checks if this action was just released.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/input/action.hpp:66]: Same as above

File: engine/include/cubos/engine/input/axis.hpp [hunk 4]
```diff
@@ -34,7 +34,7 @@ namespace cubos::engine
         /// @param positive Positive key bindings.
         /// @param negative Negative key bindings.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/input/axis.hpp:35]:
```suggestion
        /// @param positive Positive input combinations.
        /// @param negative Negative input combinations.
```

File: engine/include/cubos/engine/input/axis.hpp [hunk 5]
```diff
@@ -44,23 +44,23 @@ namespace cubos::engine
 
         /// @brief Gets the positive key bindings.
         /// @return Vector of positive keys.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/input/axis.hpp:46]:
```suggestion
        /// @brief Gets the positive input combinations.
        /// @return Vector of positive input combinations.
```

File: engine/include/cubos/engine/input/axis.hpp [hunk 6]
```diff
@@ -44,23 +44,23 @@ namespace cubos::engine
...
-        const std::vector<core::io::KeyWithModifiers>& positive() const;
+        const std::vector<InputCombination>& positive() const;
 
         /// @brief Gets the negative key bindings.
         /// @return Vector of negative keys.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/input/axis.hpp:50]: Ditto

File: engine/include/cubos/engine/input/axis.hpp [hunk 7]
```diff
@@ -44,23 +44,23 @@ namespace cubos::engine
...
-        const std::vector<core::io::KeyWithModifiers>& negative() const;
+        const std::vector<InputCombination>& negative() const;
 
         /// @brief Gets the gamepad axis bindings.
         /// @return Vector of gamepad axes.
         const std::vector<core::io::GamepadAxis>& gamepadAxes() const;
 
         /// @brief Gets the positive key bindings.
         /// @return Vector of positive keys.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/input/axis.hpp:58]: Ditto

File: engine/include/cubos/engine/input/axis.hpp [hunk 8]
```diff
@@ -44,23 +44,23 @@ namespace cubos::engine
...
-        std::vector<core::io::KeyWithModifiers>& positive();
+        std::vector<InputCombination>& positive();
 
         /// @brief Gets the negative key bindings.
         /// @return Vector of negative keys.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/input/axis.hpp:62]: Ditto

File: engine/include/cubos/engine/input/combination.hpp [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        /// @return Vector of buttons.
+        const std::vector<core::io::MouseButton>& mouseButtons() const;
+
+        /// @brief Gets the mouse buttons to be pressed.
+        /// @return Vector of buttons.
+        std::vector<core::io::MouseButton>& mouseButtons();
+
+        /// @brief Checks if the key combination is pressed.
+        /// @param window The window to check for key and mouse button presses.
+        /// @param gamepad The gamepad to check for button presses.
```
Comment (RiscadoA) [inline engine/include/cubos/engine/input/combination.hpp:71]:
```suggestion
        /// @param window Window to check for key and mouse button presses.
        /// @param gamepad Gamepad to check for button presses.
```

File: engine/samples/input/assets/sample.bind
Comment (RiscadoA) [inline engine/samples/input/assets/sample.bind:1]: I'm assuming you didn't use a formatter in these .bind files. I agree it is more readable this way, but can we enforce it?

**Review by RiscadoA — APPROVED** [review]
Thanks for working on this! You still need to rebase this, since it was based on the flake PR.
Will add a commit fixing the clang-tidy CI thing.

**Comment by diogomsmiranda** [comment]
I believe Ricardo already approved this on behalf of the team, although we should start normalizing incorporating tests to the each feature related PR we do. It would not only help with reviewing the PR but also help with the lack of current tests we have.

cc @RiscadoA

**Review by diogomsmiranda — APPROVED** [review]

**Review by Dacops — APPROVED** [review]
Post resolving the still active conversations lgtm

File: engine/samples/input/assets/sample.bind
Comment (luishfonseca) [reply engine/samples/input/assets/sample.bind:1]: Probably not, but at least for the input sample I think readability trumps format consistency

**Comment by luishfonseca** [comment]
> I believe Ricardo already approved this on behalf of the team, however we should start normalizing incorporating tests to the each feature related PR we do. It would not only help with reviewing the PR but also help with the lack of current tests we have.

I'll be rewriting a lot of the input plugin in one of my next PRs, I'll add tests then. Don't want to write them now as some functionality will change
